shiny/_main_create.py

```diff
@@ -37,40 +38,173 @@
 back_choice: Choice = Choice(title=[("class:secondary", "← Back")], value="back")
 
 
-def choice_from_dict(choice_dict: dict[str, str]) -> list[Choice]:
-    return [Choice(title=key, value=value) for key, value in choice_dict.items()]
+def choice_from_templates(templates: list[ShinyTemplate]) -> list[Choice]:
+    return [Choice(title=t.title, value=t.name) for t in templates]
+
+
+@dataclass
+class ShinyTemplate:
+    """
+    Shiny Template
+
+    This class receives metadata for a Shiny template from a `_template.json` file.
+    (Alternatively, a template can be created from just `name` and `path` for legacy
+    reasons.)
+
+    Attributes
+    ----------
+    name
+        The name of the Shiny template.
+    path
+        The path to the `_template.json` file or the root directory of the template.
+    type
+        The type of the template (e.g. "app", "package"), default: 'app'.
+    title
+        A brief title for the template, if provided.
+    description
+        A longer description of the template, if provided.
+    next_steps
+        A list of next steps or instructions related to this template, shown after the
+        default instructions are displayed. In the `_template.json` file, this field
+        can be a single string or an array of strings.
+    follow_up
+        A list of follow-up actions or information related to this template. In the
+        `_template.json` file, this field can be a single string, an array of strings,
+        or an object with a `text` field and an optional `type` field. The `type` field
+        can be "action", "info", "warning", "danger", or "text".
+    """
+
+    name: str
+    path: Path
+    type: str = "app"
+    title: str | None = None
+    description: str | None = None
+    next_steps: list[str] = field(default_factory=list)
+    follow_up: list[ShinyTemplateFollowUp] = field(default_factory=list)
+    _express_available: bool | None = None
+
+    @property
+    def express_available(self) -> bool:
+        """
+        Does the template include an Express variant, denoted by the presence of an
+        `app-express.py` file?
+        """
+
+        if self._express_available is None:
+            self._express_available = (self.path / "app-express.py").exists()
+        return self._express_available
+
+
+class ShinyTemplateFollowUp:
+    def __init__(
+        self,
+        text: str,
+        type: str = "text",
+    ):
+        self.text = text
+        self.type: Literal["action", "info", "warning", "danger", "text"] = "text"
+        if type in ("action", "info", "warning", "danger"):
+            self.type = type
+
+
+def find_templates(path: Path | str = ".") -> list[ShinyTemplate]:
+    path = Path(path)
+    templates: list[ShinyTemplate] = []
+
+    template_files = sorted(path.glob("**/_template.json"))
+    for tf in template_files:
+        with tf.open() as f:
+            template = json.load(f)
```

Comment:
It might be worth catching and throwing errors here? Most importantly, it seems like if an error were to occur, you wouldn't know what file (`f`) is problematic?

Reply:
`json.load()` doesn't tell you what file it failed to parse? I'll look into this

Reply:
`json.load()` _does not_ tell you much.